modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DataNodesManager.java

```diff
@@ -0,0 +1,1562 @@
+                zoneDescriptor
+        );
+
+        DistributionZoneTimer scaleUpTimerToSave = timerToSave(timestamp, modifiedScaleUpTimer);
+        DistributionZoneTimer scaleDownTimerToSave = timerToSave(timestamp, modifiedScaleDownTimer);
+
+        Condition condition = and(
+                dataNodesHistoryEqualToOrNotExists(zoneId, dataNodesHistory),
+                and(
+                        timerEqualToOrNotExists(zoneScaleUpTimerKey(zoneId), scaleUpTimer),
+                        timerEqualToOrNotExists(zoneScaleDownTimerKey(zoneId), scaleDownTimer)
+                )
+        );
+
+        Operations operations = ops(
+                addNewEntryToDataNodesHistory(zoneId, dataNodesHistory, currentDataNodes.timestamp(),
+                        currentDataNodes.dataNodes()),
+                renewTimer(zoneScaleUpTimerKey(zoneId), scaleUpTimerToSave),
+                renewTimer(zoneScaleDownTimerKey(zoneId), scaleDownTimerToSave)
+        );
+
+        return DataNodesHistoryMetaStorageOperation.builder()
+                .zoneId(zoneId)
+                .condition(condition)
+                .operations(operations)
+                .operationName("distribution zone auto adjust change")
+                .currentDataNodesHistory(dataNodesHistory)
+                .currentTimestamp(timestamp)
+                .historyEntryTimestamp(currentDataNodes.timestamp())
+                .historyEntryNodes(currentDataNodes.dataNodes())
+                .scaleUpTimer(scaleUpTimerToSave)
+                .scaleDownTimer(scaleDownTimerToSave)
+                .build();
+    }
+
+    void onUpdatePartitionDistributionReset(
+            int zoneId,
+            int partitionDistributionResetTimeoutSeconds,
+            Runnable taskOnReset
+    ) {
+        if (partitionDistributionResetTimeoutSeconds == INFINITE_TIMER_VALUE) {
+            zoneTimers.computeIfAbsent(zoneId, this::createZoneTimers).partitionReset.stopScheduledTask();
+        } else {
+            zoneTimers.computeIfAbsent(zoneId, this::createZoneTimers).partitionReset
+                    .reschedule(partitionDistributionResetTimeoutSeconds, taskOnReset);
+        }
+    }
+
+    /**
+     * The closure that is executed on scale up or scale down schedule. Appends data nodes history in meta storage and sets the timer
+     * to default value.
+     *
+     * @param zoneDescriptor Zone descriptor.
+     * @param scheduledTimer Scheduled timer.
+     * @return Runnable that is executed on schedule.
+     */
+    private Runnable applyTimerClosure(CatalogZoneDescriptor zoneDescriptor, ScheduledTimer scheduledTimer) {
+        int zoneId = zoneDescriptor.id();
+
+        return () -> doOperation(
+                zoneDescriptor,
+                List.of(zoneDataNodesHistoryKey(zoneId), scheduledTimer.metaStorageKey()),
+                dataNodesHistoryContext -> applyTimerClosure0(zoneDescriptor, scheduledTimer, dataNodesHistoryContext)
+        );
+    }
+
+    @Nullable
```

Comment:
I don't think this method is `@Nullable`

Reply:
removed